### PR TITLE
[deconz] Changed scheduler policy to fix OutOfMemoryError

### PR DESCRIPTION
**Pull request by cweitkamp**
- Changed scheduler policy

I do not trust the recursion of the `scheduler.scheduleWithFixedDelay(this::startWebsocket, ...)` calls. I think a one time shot should be sufficient. On the weekend my test environment crashed because of a `java.lang.OutOfMemoryError: Java heap space` error - If I remember correctly I updated my FRITZ!Box to the latest OS version and there was a temporarily downtime of my network:

grafik
Screenshot of the heap dump in Eclipse Memory Analyzer Tool (MAT).

Signed-off-by: Christoph Weitkamp <github@christophweitkamp.de>

### CODE REVIEW

**Comment by davidgraeff** [comment]
Hm. I'm using this pattern not only here. Good to know that this API cannot be trusted.

**Comment by wborn** [comment]
There were some OOM issues with this binding before see #4361, #4365. Maybe having a look at those helps to eliminate these once and for all.

**Comment by davidgraeff** [comment]
An ever increasing list is not something that you can easily spot, but it is not related to this issue here.

The scheduler API was correctly used, but seem to leak internally.

**Comment by wborn** [comment]
I linked those issues as reference for users stumbling on those issues still having OOMs. ;-)

**Comment by cweitkamp** [comment]
I guess issue [JDK-8206958](https://bugs.openjdk.java.net/browse/JDK-8206958) fits for our situation. The author talks about:

> I observed similiar behaviour with ScheduledExecutorService.

File: bundles/org.openhab.binding.deconz/src/main/java/org/openhab/binding/deconz/internal/handler/DeconzBridgeHandler.java [hunk 1]
```diff
@@ -98,9 +98,8 @@ public void handleCommand(ChannelUID channelUID, Command command) {
      * Stops the API request or websocket reconnect timer
      */
     private void stopTimer() {
-        ScheduledFuture<?> future = scheduledFuture;
```
Comment (martinvw) [inline bundles/org.openhab.binding.deconz/src/main/java/org/openhab/binding/deconz/internal/handler/DeconzBridgeHandler.java]: Why do you have away from the local variable concept here? I personally think it makes sense

Comment (cweitkamp) [reply bundles/org.openhab.binding.deconz/src/main/java/org/openhab/binding/deconz/internal/handler/DeconzBridgeHandler.java]: No specific one. I personally do not like this pattern. It only omits the warning in Eclipse IDE. Or did I miss something. But I do not have a problem with reverting it if you want me to do that.

Comment (J-N-K) [reply bundles/org.openhab.binding.deconz/src/main/java/org/openhab/binding/deconz/internal/handler/DeconzBridgeHandler.java]: From my understanding this keeps the reference even if some other method modifies the field (or sets it to null), so we make sure that we always act on the object that was valid when we assigned the local variable.

Comment (cweitkamp) [reply bundles/org.openhab.binding.deconz/src/main/java/org/openhab/binding/deconz/internal/handler/DeconzBridgeHandler.java]: Good point. Does this mean we will watch out for it and tell people to use it in future PRs?

Comment (wborn) [reply bundles/org.openhab.binding.deconz/src/main/java/org/openhab/binding/deconz/internal/handler/DeconzBridgeHandler.java]: That's right! There's a small possibility the variable is just reassigned after doing the null check. So it's never safe to reference fields unless access is synchronized. So the pattern is there to fix bugs and not to please the IDE. :wink:

Comment (wborn) [reply bundles/org.openhab.binding.deconz/src/main/java/org/openhab/binding/deconz/internal/handler/DeconzBridgeHandler.java]: The pattern is also advocated by the Eclipse null annotations [documentation (The case of fields)](https://help.eclipse.org/oxygen/index.jsp?topic=%2Forg.eclipse.jdt.doc.user%2Ftasks%2Ftask-using_null_annotations.htm):

> The situation is more delicate when considering a field marked `@Nullable`. Such a field should always be considered as dangerous and the recommended way of working with nullable fields is: always assign the value to a local variable before working with it. Using a local variable the flow analysis can tell exactly, whether a dereference is sufficiently protected by a null check. When following this general rule working with nullable fields poses no problems.

Comment (cweitkamp) [reply bundles/org.openhab.binding.deconz/src/main/java/org/openhab/binding/deconz/internal/handler/DeconzBridgeHandler.java]: Thanks. I reverted it.

**Pull request merged**